File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/models/DomainModelsAreImmutable.kt
package com.perrystreet.woof.konsist.models

import com.lemonappdev.konsist.api.ext.list.properties
import com.perrystreet.woof.konsist.Assertions.assertFalse
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class DomainModelsAreImmutable : BehaviorSpec() {
    init {
        Given("A property of a domain model") {
            val properties = KonsistUtils.domainModelModule.classes().properties()

            Then("It is declared with val") {
                properties.assertFalse(message = Message) { it.hasVarModifier }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "Domain models only have val properties.",
            why = "Domain models flow through RxJava streams shared by several ViewModels. Mutable fields would let one consumer change what another is observing.",
            howToFix = "Use val and copy() to derive new instances.",
            badExample = "data class Dog(var name: String)",
            goodExample = "data class Dog(val name: String)",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/Models/DomainModelsAreImmutable.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class DomainModelsAreImmutable: QuickSpec {
    override class func spec() {
        Given("A stored property of a domain model") {
            let properties = WoofHarmonize.domainModelPackage.structs().flatMap(\.variables).filter(\.isStored)

            Then("It is declared with let") {
                properties.assertTrue(rule: rule) { $0.isConstant }
            }
        }
    }

    private static let rule = Rule(
        description: "Domain models only have let properties.",
        rationale: "Domain models flow through Combine streams shared by several ViewModels. Mutable fields would let one consumer change what another is observing.",
        fixHint: "Use let and build new instances to derive changes.",
        badExample: "struct Dog { var name: String }",
        goodExample: "struct Dog { let name: String }"
    )
}
